File: model/src/main/java/net/artux/pda/model/QuestUtil.java
package net.artux.pda.model;

import net.artux.pda.model.items.ItemModel;
import net.artux.pda.model.quest.story.StoryDataModel;

import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

public class QuestUtil {

    public static boolean check(HashMap<String, List<String>> conditions, StoryDataModel storyDataModel) {
        if (conditions != null) {
            if (!conditions.keySet().isEmpty()) {
                HashMap<String, Integer> map = storyDataModel.getParametersMap();
                for (String condition : conditions.keySet()) {
                    switch (condition) {
                        case "has":
                            if (conditions.get(condition) != null)
                                for (String has : conditions.get(condition)) {
                                    if (!map.containsKey(has)
                                            && !containsItem(storyDataModel.getAllItems(), has))
                                        return false;
                                }
                            break;
                        case "!has":
                            if (conditions.get(condition) != null)
                                for (String has : conditions.get(condition)) {
                                    if (map.containsKey(has)
                                            || containsItem(storyDataModel.getAllItems(), has))
                                        return false;

                                }
                            break;
                        case "money>=":
                            if (conditions.get(condition) != null)
                                for (String money : conditions.get(condition)) {
                                    if (storyDataModel.getMoney() < Integer.parseInt(money))
                                        return false;
                                }
                            break;
                        case "money<":
                            if (conditions.get(condition) != null)
                                for (String money : conditions.get(condition)) {
                                    if (storyDataModel.getMoney() >= Integer.parseInt(money))
                                        return false;
                                }
                            break;

                    }
                }
            }
        }
        return true;
    }

    private static boolean containsItem(List<ItemModel> items, String sid) {
        if (isInteger(sid)) {
            int id = Integer.parseInt(sid);
            for (ItemModel item : items) {
                if (item.getBaseId() == id)
                    return true;
            }
        }

        return false;
    }


    public static boolean isInteger(String str) {
        if (str == null) {
            return false;
        }
        int length = str.length();
        if (length == 0) {
            return false;
        }
        int i = 0;
        if (str.charAt(0) == '-') {
            if (length == 1) {
                return false;
            }
            i = 1;
        }
        for (; i < length; i++) {
            char c = str.charAt(i);
            if (c < '0' || c > '9') {
                return false;
            }
        }
        return true;
    }

    public static Map<String, List<String>> difference(StoryDataModel oldData, StoryDataModel newData) {
        List<ItemModel> oldItems = oldData.getAllItems();
        List<String> itemDifferences = new LinkedList<>();
        List<String> newItems = newData.getAllItems()
                .stream()
                .filter(itemModel -> itemModel.getId() == null)
                .map(itemModel -> itemModel.getBaseId() + ":" + itemModel.getQuantity())
                .collect(Collectors.toList());

        HashMap<String, List<String>> response = new HashMap<>(Collections.singletonMap("add_items", newItems));

        for (ItemModel newItem : newData.getAllItems()) {
            Optional<ItemModel> difference = oldItems.stream()
                    .filter(itemModel -> itemModel.getId() != null && itemModel.getId().equals(newItem.getId()) &&
                            itemModel.getQuantity() != newItem.getQuantity())
                    .findFirst();
            difference.ifPresent(itemModel -> itemDifferences.add(newItem.getId() + ":" + newItem.getQuantity()));
        }
        response.put("item", itemDifferences);

        return response;
    }

}

File: model/src/main/java/net/artux/pda/model/QuestUtil.kt
package net.artux.pda.model

import net.artux.pda.model.items.ItemModel
import net.artux.pda.model.items.WearableModel
import net.artux.pda.model.quest.story.StoryDataModel
import java.util.LinkedList
import java.util.stream.Collectors

object QuestUtil {

    @JvmStatic
    fun check(conditions: Map<String, List<String>>?, storyDataModel: StoryDataModel): Boolean {
        if (conditions.isNullOrEmpty())
            return true
        val map = storyDataModel.parametersMap
        map["money"] = storyDataModel.money
        map["xp"] = storyDataModel.xp

        for (key in conditions.keys) {
            if (conditions[key].isNullOrEmpty())
                continue

            when (key) {
                "has" -> for (param in conditions[key]!!) {
                    if (!map.containsKey(param)
                        && !containsItem(storyDataModel.allItems, param)
                    ) return false
                }

                "!has" -> for (param in conditions[key]!!) {
                    if (map.containsKey(param)
                        || containsItem(storyDataModel.allItems, param)
                    ) return false
                }

                else -> {
                    // split("<", "<=", ">=", ">", "=", "!=")
                    val userParam = key.filter { it.isLetter() }
                    val operation = key.filter { !it.isLetterOrDigit() }
                    for (param in conditions[key]!!){
                        if(!checkParam(userParam, operation, param.toInt(), map))
                            return false
                    }

                }
            }
        }
        return true
    }

    private fun checkParam(param: String, operation: String, conditionValue: Int, parameterMap: Map<String, Int>): Boolean{
        val currentValue = parameterMap[param] ?: return false
        println("$param - ${currentValue} $operation $conditionValue")
        when (operation){
            "<" -> if (currentValue >= conditionValue) return false
            ">" -> if (currentValue <= conditionValue) return false
            ">=" -> if (currentValue < conditionValue) return false
            "<=" -> if (currentValue > conditionValue) return false
            "===","==","=" -> if (currentValue != conditionValue) return false
            "!=" -> if (currentValue == conditionValue) return false
        }
        return true
    }

    private fun containsItem(items: List<ItemModel>, sid: String): Boolean {
        if (isInteger(sid)) {
            val id = sid.toInt()
            for (item in items) {
                if (item.baseId == id) return true
            }
        }
        return false
    }

    private fun isInteger(str: String?): Boolean {
        if (str == null) {
            return false
        }
        val length = str.length
        if (length == 0) {
            return false
        }
        var i = 0
        if (str[0] == '-') {
            if (length == 1) {
                return false
            }
            i = 1
        }
        while (i < length) {
            val c = str[i]
            if (c < '0' || c > '9') {
                return false
            }
            i++
        }
        return true
    }

    fun calculateDifference(oldData: StoryDataModel, newData: StoryDataModel): LinkedHashMap<String, List<String>> {
        val response = LinkedHashMap<String, List<String>>()

        val oldItems = oldData.allItems
        val itemDifferences: MutableList<String> = LinkedList()
        //new items
        val newItems = newData.allItems
            .stream()
            .filter { itemModel: ItemModel -> itemModel.id == null }
            .map { itemModel: ItemModel -> itemModel.baseId.toString() + ":" + itemModel.quantity }
            .collect(Collectors.toList())
        itemDifferences.addAll(newItems)


        // quantity and quality difference
        for (newItem in newData.allItems) {
            val quantityDifference = oldItems.stream()
                .filter { itemModel: ItemModel ->
                    itemModel.id != null
                            && itemModel.id == newItem.id
                            && itemModel.quantity != newItem.quantity
                }
                .findFirst()
            quantityDifference.ifPresent { itemDifferences.add(newItem.id.toString() + ":" + newItem.quantity) }
        }
        response["item"] = itemDifferences

        // money difference
        response["money"] = mutableListOf((newData.money - oldData.money).toString())

        //xp difference
        response["xp"] = mutableListOf((newData.xp - oldData.xp).toString())

        // wearable difference
        val wearableItems = mutableListOf<String>()
        newData.allItems.stream()
            .filter { itemModel: ItemModel -> itemModel.type.isWearable }
            .map { it: ItemModel -> it as WearableModel }
            .forEach {
                if (it.isEquipped)
                    if (it.id == null) {
                        wearableItems.add(it.baseId.toString())
                    } else {
                        wearableItems.add(it.id.toString())
                    }
            }

        oldItems.stream()
            .filter { itemModel: ItemModel -> itemModel.type.isWearable }
            .map { it: ItemModel -> it as WearableModel }
            .forEach { it ->
                if (it.isEquipped) {
                    val id = it.id.toString()
                    wearableItems.removeIf { it == id }
                }
            }
        response["set"] = wearableItems

        return response
    }
}